src/libraries/System.Formats.Tar/src/System/Formats/Tar/TarHeader.Read.cs

```diff
@@ -28,45 +30,162 @@ internal bool TryGetNextHeader(Stream archiveStream, bool copyData)
 
             archiveStream.ReadExactly(buffer);
 
-            try
+            // Confirms if v7 or pax, or tentatively selects ustar
+            if (!TryReadCommonAttributes(buffer))
             {
-                // Confirms if v7 or pax, or tentatively selects ustar
-                if (!TryReadCommonAttributes(buffer))
+                return false;
+            }
+
+            // Confirms if gnu, or tentatively selects ustar
+            ReadMagicAttribute(buffer);
+
+            if (_format != TarEntryFormat.V7)
+            {
+                // Confirms if gnu
+                ReadVersionAttribute(buffer);
+
+                // Fields that ustar, pax and gnu share identically
+                ReadPosixAndGnuSharedAttributes(buffer);
+
+                Debug.Assert(_format is TarEntryFormat.Ustar or TarEntryFormat.Pax or TarEntryFormat.Gnu);
+                if (_format == TarEntryFormat.Ustar)
+                {
+                    ReadUstarAttributes(buffer);
+                }
+                else if (_format == TarEntryFormat.Gnu)
                 {
-                    return false;
+                    ReadGnuAttributes(buffer);
                 }
+                // In PAX, there is nothing to read in this section (empty space)
+            }
+
+            ProcessDataBlock(archiveStream, copyData);
+
+            ArrayPool<byte>.Shared.Return(rented);
```

Comment:
since `RecordSize` is quite small (512 bytes), there is no need to use `ArrayPool`, we can just stack allocate the buffer